packages/effect/src/Effect.ts

```diff
@@ -6305,3 +6325,309 @@ export const Tag: <const Id extends string>(id: Id) => <
+    } ?
+        & Return<Self, Id, Type, Proxy>
+        & Layer.Layer<
+          Self,
+          { [k in keyof Layers]: Layer.Layer.Error<Layers[k]> }[number],
+          { [k in keyof Layers]: Layer.Layer.Context<Layers[k]> }[number]
+        >
+        & {
+          readonly Layer: Layer.Layer<Self, never, never>
+        }
+    : Maker extends {
+      sync: () => infer Type
+      dependencies: []
+    } ? Service.Return<Self, Id, Type, Proxy> & Layer.Layer<Self, never, never> & {
+        readonly Layer: Layer.Layer<Self, never, never>
+      }
+    : Maker extends {
+      sync: () => infer Type
+    } ? Return<Self, Id, Type, Proxy> & Layer.Layer<Self, never, never> & {
+        readonly Layer: Layer.Layer<Self, never, never>
+      }
+    : never
+}
+
+/**
+ * @since 2.0.0
+ * @category constructors
+ */
+export const Service: {
+  <Self>(): {
+    <
+      const Id extends string,
+      Maker extends (Service.Maker<Service.ProhibitedType> & { proxy: true }) | (Service.Maker<{}> & { proxy?: false })
+    >(
+      id: Id,
+      maker: Maker
+    ): Service.ReturnWithMaker<Self, Id, Maker, Maker extends { proxy: true } ? true : false>
+  }
+} = function() {
+  return function() {
+    const [id, maker] = arguments
+    const proxy = "proxy" in maker ? maker["proxy"] : false
+    const limit = Error.stackTraceLimit
+    Error.stackTraceLimit = 2
+    const creationError = new Error()
+    Error.stackTraceLimit = limit
+    function TagClass() {}
+    Object.setPrototypeOf(TagClass, TagProto)
+    TagClass.key = id
+    Object.defineProperty(TagClass, "stack", {
+      get() {
+        return creationError.stack
+      }
+    })
+
+    TagClass["of"] = (args: any) =>
+      new Proxy(args, {
+        get(_target: any, prop: any, _receiver) {
+          if (prop === "_tag") {
+            return id
+          }
+          return args[prop]
+        }
+      })
+
+    const of = TagClass["of"]
+
+    if ("effect" in maker) {
+      // @ts-expect-error
+      TagClass["Layer"] = layer.fromEffect(TagClass, maker["effect"]).pipe(
+        // @ts-expect-error
+        layer.flatMap((c) => layer.succeedContext(Context.add(c, TagClass, of(Context.unsafeGet(c, TagClass)))))
+      )
+      let live = TagClass["Layer"]
+      if ("dependencies" in maker) {
+        live = live.pipe(layer.provide(layer.mergeAll(...maker["dependencies"])))
+      }
+      Object.assign(TagClass, live)
+      Object.assign(TagClass, layer.proto)
+    }
+
+    if ("scoped" in maker) {
+      // @ts-expect-error
+      TagClass["Layer"] = layer.scoped(TagClass, maker["scoped"]).pipe(
+        // @ts-expect-error
+        layer.flatMap((c) => layer.succeedContext(Context.add(c, TagClass, of(Context.unsafeGet(c, TagClass)))))
+      )
+      let live = TagClass["Layer"]
+      if ("dependencies" in maker) {
+        live = live.pipe(layer.provide(layer.mergeAll(...maker["dependencies"])))
```

Comment:
Could we make Layer.provide variadric aswell?

Reply:
Yes